nixos/modules/system/boot/initrd-ssh.nix

```diff
@@ -10,123 +10,192 @@ in
+    '') [ "hostRSAKey" "hostDSSKey" "hostECDSAKey" ];
+
+  config = let
+    initrdKeyPath = k:
+      if k ? path && isString k.path
+        then k.path
+        else
+          # Nix complains if you include a store hash in initrd path
+          # names, so here's an awful hack. It also helps us ensure
+          # uniqueness for keys without a path set.
+          let hash = builtins.hashString "sha256"
+            (builtins.unsafeDiscardStringContext (builtins.toJSON k));
+          in "/etc/ssh/ssh_host_${k.type}_key_${hash}";
+
+    sshdCfg = config.services.openssh;
+
+    sshdConfig = ''
+      Port ${toString cfg.port}
+
+      PasswordAuthentication no
+      ChallengeResponseAuthentication no
+
+      ${flip concatMapStrings cfg.hostKeys (k: ''
+        HostKey ${initrdKeyPath k}
+      '')}
+
+      KexAlgorithms ${concatStringsSep "," sshdCfg.kexAlgorithms}
+      Ciphers ${concatStringsSep "," sshdCfg.ciphers}
+      MACs ${concatStringsSep "," sshdCfg.macs}
+
+      LogLevel ${sshdCfg.logLevel}
+
+      ${if sshdCfg.useDns then ''
+        UseDNS yes
+      '' else ''
+        UseDNS no
+      ''}
+    '';
+  in mkIf (config.boot.initrd.network.enable && cfg.enable) {
     assertions = [
       { assertion = cfg.authorizedKeys != [];
         message = "You should specify at least one authorized key for initrd SSH";
       }
     ];
 
     boot.initrd.extraUtilsCommands = ''
-      copy_bin_and_libs ${pkgs.dropbear}/bin/dropbear
+      copy_bin_and_libs ${pkgs.openssh}/bin/sshd
+      copy_bin_and_libs ${pkgs.openssh}/bin/ssh-keygen
       cp -pv ${pkgs.glibc.out}/lib/libnss_files.so.* $out/lib
     '';
 
     boot.initrd.extraUtilsCommandsTest = ''
-      $out/bin/dropbear -V
+      # sshd requires a host key to check config, so we pass in the test's
+      echo -n ${escapeShellArg sshdConfig} |
+        $out/bin/sshd -t -f /dev/stdin \
+        -h ${../../../tests/initrd-network-ssh/ssh_host_ed25519_key}
     '';
 
     boot.initrd.network.postCommands = ''
       echo '${cfg.shell}' > /etc/shells
       echo 'root:x:0:0:root:/root:${cfg.shell}' > /etc/passwd
+      echo 'sshd:x:1:1:sshd:/var/empty:/bin/nologin' >> /etc/passwd
       echo 'passwd: files' > /etc/nsswitch.conf
 
-      mkdir -p /var/log
+      mkdir -p /var/log /var/empty
       touch /var/log/lastlog
 
-      mkdir -p /etc/dropbear
+      mkdir -p /etc/ssh
+      echo -n ${escapeShellArg sshdConfig} > /etc/ssh/sshd_config
+
+      echo "export PATH=$PATH" >> /etc/profile
+      echo "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH" >> /etc/profile
 
       mkdir -p /root/.ssh
       ${concatStrings (map (key: ''
         echo ${escapeShellArg key} >> /root/.ssh/authorized_keys
       '') cfg.authorizedKeys)}
 
-      dropbear -s -j -k -E -p ${toString cfg.port} ${optionalString (cfg.hostRSAKey == null && cfg.hostDSSKey == null && cfg.hostECDSAKey == null) "-R"}
+      ${flip concatMapStrings cfg.hostKeys (k: ''
+        if [ -f "${initrdKeyPath k}" ]; then
```

Comment:
This condition is known at nix evaluation time, isn't it?
There is no way to supply secrets to initrd, around the creation of initrd.

Thus, whether `ssh-keygen` be included into `initrd` is known at nix evaluation time, and it might be excluded if not needed.

Reply:
The increase in initrd size from including `ssh-keygen` is only 196 KiB, so I don't think the complexity of the additional Nix logic that would be required is worthwhile.

Reply:
How is it more complex than a bash conditional?

Reply:
To clarify, I was referring to

> Thus, whether `ssh-keygen` be included into `initrd` is known at nix evaluation time, and it might be excluded if not needed.

which didn't seem worth the required additional Nix logic to me. (But again, I'm happy to just remove the `ssh-keygen` support if people are okay breaking compatibility with existing configurations that don't set any host keys.)

Reply:
I think we can just make it mandatory to have at least one key.
Regenerating keys on each boot does not provide a lot security over using telnet since a potential attacker in the same network could come up with its own server.